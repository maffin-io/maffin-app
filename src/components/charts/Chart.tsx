'use client';

import React from 'react';
import dynamic from 'next/dynamic';
import { ApexOptions } from 'apexcharts';

import { toFixed } from '@/helpers/number';

export type ChartProps = {
  type: 'line' | 'bar' | 'pie' | 'donut',
  series?: ApexOptions['series'],
  labels?: string[],
  title?: string,
  showLegend?: boolean,
  xCategories?: string[],
  hideSeries?: string[],
  unit?: string,
  height?: number,
  xAxisType?: 'datetime' | 'category' | 'numeric' | undefined,
};

// apexcharts import references window so we need this
// https://stackoverflow.com/questions/68596778
// However, I haven't found a neat way to fix types that's why
// we need the ts-ignore in the return.
const ApexChart = dynamic(
  () => import('react-apexcharts').then((module) => module.default),
  {
    ssr: false,
    loading: () => <p>Loading...</p>,
  },
);

export default function Chart({
  type,
  title = '',
  series = [],
  labels = [],
  showLegend = true,
  xCategories = [],
  hideSeries = [],
  unit = '',
  height = 400,
  xAxisType = undefined,
}: ChartProps): JSX.Element {
  if (!series?.length) {
    return (
      <span>Loading...</span>
    );
  }

  let options = OPTIONS;
  options = {
    ...options,
    labels,
    chart: {
      ...options.chart,
      id: title,
      events: {
        mounted: (chart) => hideSeries.forEach(name => {
          try {
            chart.hideSeries(name);
          } catch {
            // this fails sometimes for some reason but still renders
            // as expected. Adding the catch to protect against that.
          }
        }),
      },
    },
    legend: {
      show: showLegend,
    },
    title: {
      text: title,
      align: 'left',
    },
    xaxis: {
      ...options.xaxis,
      categories: xCategories,
      type: xAxisType,
    },
    yaxis: {
      labels: {
        formatter: (val: number) => `${toFixed(val)} ${unit}`,
      },
    },
    tooltip: {
      ...options.tooltip,
      y: {
        formatter: (val: number) => `${toFixed(val)} ${unit}`,
      },
    },
    stroke: {
      ...options.stroke,
      width: type === 'line' ? 1 : 0,
      dashArray: type === 'line' ? 5 : 0,
    },
  };

  return (
    // @ts-ignore
    <ApexChart
      options={options}
      series={series}
      type={type}
      className="apex-charts"
      // https://stackoverflow.com/questions/75103994
      width="100%"
      height={height}
    />
  );
}

const OPTIONS: ApexOptions = {
  grid: {
    borderColor: '#777f85',
  },
  chart: {
    id: 'line',
    width: '100%',
    foreColor: '#94A3B8',
    toolbar: {
      show: false,
    },
    zoom: {
      type: 'x',
      enabled: true,
      autoScaleYaxis: true,
    },
  },
  stroke: {
    dashArray: 5,
    curve: 'smooth',
  },
  xaxis: {
    axisBorder: {
      show: false,
    },
  },
  plotOptions: {
    bar: {
      horizontal: false,
      columnWidth: '55%',
    },
  },
  dataLabels: {
    enabled: false,
  },
  tooltip: {
    fillSeriesColor: true,
    x: {
      show: false,
    },
    theme: 'dark',
    intersect: true,
    inverseOrder: true,
    shared: false,
  },
};
