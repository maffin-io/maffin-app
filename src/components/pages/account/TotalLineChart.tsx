import React from 'react';
import { ApexOptions } from 'apexcharts';

import Chart from '@/components/charts/Chart';
import { Split } from '@/book/entities';
import { currencyToSymbol } from '@/book/helpers';

export type TotalLineChartProps = {
  splits: Split[],
};

export default function TotalLineChart({
  splits,
}: TotalLineChartProps): JSX.Element {
  const series: ApexOptions['series'] = [{ data: [] }];

  let totalAggregate = 0;
  splits.slice().reverse().forEach(split => {
    let { quantity } = split;
    if (split.account.type === 'INCOME') {
      quantity = -quantity;
    }
    totalAggregate += quantity;

    (series[0].data as { x: number, y: number }[]).push({
      x: split.transaction.date.toMillis(),
      y: totalAggregate,
    });
  });

  return (
    <div className="bg-gunmetal-700 rounded-sm my-6 mr-6">
      <Chart
        type="line"
        series={series}
        unit={currencyToSymbol(splits[0]?.account.commodity.mnemonic || '')}
        height={255}
        xAxisType="datetime"
      />
    </div>
  );
}
