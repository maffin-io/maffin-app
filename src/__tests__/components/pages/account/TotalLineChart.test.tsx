import React from 'react';
import { DateTime } from 'luxon';
import { render } from '@testing-library/react';

import type { Split } from '@/book/entities';
import Chart from '@/components/charts/Chart';
import TotalLineChart from '@/components/pages/account/TotalLineChart';

jest.mock('@/components/charts/Chart', () => jest.fn(
  () => <div data-testid="Chart" />,
));

describe('TotalLineChart', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('creates Chart with expected params', () => {
    render(<TotalLineChart splits={[]} />);

    expect(Chart).toBeCalledWith(
      {
        height: 255,
        series: [{ data: [] }],
        type: 'line',
        unit: undefined,
        options: {
          chart: {
            zoom: {
              enabled: false,
            },
          },
          xaxis: {
            type: 'datetime',
          },
        },
      },
      {},
    );
  });

  it('builds series accumulating values', () => {
    // note that splits are received ordered already. Without order, it will compute
    // wrongly
    render(
      <TotalLineChart
        splits={[
          {
            account: {
              type: 'ASSET',
              commodity: {
                mnemonic: 'EUR',
              },
            },
            transaction: {
              date: DateTime.fromISO('2023-01-02'),
            },
            quantity: 100,
          } as Split,
          {
            account: {
              // This is not very accurate as type is always the same but
              // this way we test INCOME behavior in this same test
              type: 'INCOME',
              commodity: {
                mnemonic: 'EUR',
              },
            },
            transaction: {
              date: DateTime.fromISO('2023-01-01'),
            },
            quantity: -200,
          } as Split,
        ]}
      />,
    );

    expect(Chart).toBeCalledWith(
      {
        height: 255,
        series: [
          {
            data: [
              {
                x: 1672531200000,
                y: 200,
              },
              {
                x: 1672617600000,
                y: 300,
              },
            ],
          },
        ],
        type: 'line',
        unit: 'EUR',
        options: {
          chart: {
            zoom: {
              enabled: false,
            },
          },
          xaxis: {
            type: 'datetime',
          },
        },
      },
      {},
    );
  });
});
