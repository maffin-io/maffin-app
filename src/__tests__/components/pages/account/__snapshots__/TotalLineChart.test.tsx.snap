// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TotalLineChart creates Chart with expected params 1`] = `
<div>
  <div
    class="bg-gunmetal-700 rounded-sm my-6 mr-6"
  >
    <div
      data-testid="Chart"
    />
  </div>
</div>
`;
